test.sh

```diff
@@ -1,60 +1,73 @@
 #!/bin/bash
-set -ex
+set -ex -o pipefail
 
-docker_build() {
-  echo "Target dir: $TARGET_DIR"
-  echo "Platform: $PLATFORM"
+# Common vars:
+CRATE_NAME="${1}crate"
+CRATE_PATH="./test/${CRATE_NAME}"
+
+# Build and verify successful static compilation of a crate:
+function docker_build() {
+  echo "Target dir: ${TARGET_DIR}"
+  echo "Platform: ${PLATFORM}"
 
   # NB: add -vv to cargo build when debugging
-  local -r crate="$1"crate
-  docker run --rm \
-    -v "$PWD/test/${crate}:/volume" \
-    -v cargo-cache:/root/.cargo/registry \
-    -e RUST_BACKTRACE=1 \
+  docker run --rm -it \
+    --env RUST_BACKTRACE=1 \
+    --volume "${CRATE_PATH}:/volume" \
+    --volume cargo-cache:/opt/cargo/registry \
     --platform "${PLATFORM}" \
     rustmusl-temp \
     cargo build
 
-  cd "test/${crate}"
-
-  # Ideally we would use `ldd` but due to a qemu bug we can't :(
-  # See https://github.com/multiarch/qemu-user-static/issues/172
-  # Instead we use `file`.
-  docker run --rm \
-    -v "$PWD:/volume" \
-    -e RUST_BACKTRACE=1 \
+  # Verify the build artifact works and is statically linked:
+  # (A container is used for `ldd` so that a non-native platform can also be tested)
+  local CRATE_ARTIFACT="./target/${TARGET_DIR}/debug/${CRATE_NAME}"
+  docker run --rm -it \
+    --env RUST_BACKTRACE=1 \
+    --volume "${CRATE_PATH}:/volume" \
+    --workdir /volume \
     --platform "${PLATFORM}" \
     test-runner \
-    bash -c "cd volume; ./target/$TARGET_DIR/debug/${crate} && file ./target/$TARGET_DIR/debug/${crate} && file /volume/target/$TARGET_DIR/debug/${crate} 2>&1 | grep -qE 'static-pie linked|statically linked' && echo ${crate} is a static executable"
+    bash -ex -o pipefail -c "
```

Comment:
I added `-o pipefail` here (and at the start of the script) to ensure that if there was a non-zero exit status in `ldd` command it'd not be replaced after the pipe to `grep`. Unlikely to be relevant in practice though.

Likewise since the check is running in another shell instance of `bash`, I matched the `-ex` options for consistency with the other helpers local check method.